dev/release/setup-centos7.sh

```diff
@@ -0,0 +1,90 @@
+
+#!/bin/sh
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+
+# A script to install dependencies required for release
+# verification Red Hat Enterprise Linux 7 clones in particular
+# on CentOS 7
+
+set -exu
+
+yum -y update
+yum -y groupinstall "Development Tools"
+yum install -y centos-release-scl curl
+cd /etc/pki/rpm-gpg
+curl -O http://springdale.princeton.edu/data/springdale/7/x86_64/os/RPM-GPG-KEY-springdale
+cd /etc/yum.repos.d
+cat << EOF > Springdale-SCL.repo
+[Springdale-SCL]
+name=Springdale - SCL
+baseurl=http://springdale.princeton.edu/data/springdale/SCL/$releasever/$basearch
+gpgcheck=1
+enabled=1
+gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-springdale
+EOF
+cd $HOME
+yum -y install \
+  cmake3 \
+  devtoolset-11-* \
+  git \
+  gobject-introspection-devel \
+  java-11-openjdk-devel \
+  libcurl-devel \
+  libicu-devel \
+  libtool \
+  llvm-toolset-13.0-* \
+  maven \
+  ncurses-devel \
+  ninja-build \
+  perl-core \
+  rh-python36 \
+  sqlite-devel \
+  tar \
+  vala-devel \
+  wget \
+  which \
+  zlib-devel
+  
+source scl_source enable devtoolset-11
+source scl_source enable llvm-toolset-13.0
+git clone https://github.com/sstephenson/rbenv.git ~/.rbenv
+echo 'export PATH="$HOME/.rbenv/bin:$PATH"' >> ~/.bash_profile
+echo 'eval "$(rbenv init -)"' >> ~/.bash_profile
+source  ~/.bash_profile
+rbenv install 3.1.2
+rbenv global 3.1.2
+
+alternatives --install /usr/local/bin/cmake cmake /usr/bin/cmake3 20 \
+--slave /usr/local/bin/ctest ctest /usr/bin/ctest3 \
+--slave /usr/local/bin/cpack cpack /usr/bin/cpack3 \
+--slave /usr/local/bin/ccmake ccmake /usr/bin/ccmake3 \
+--family cmake
+scl enable rh-python36 bash
+python -m pip install -U pip
+wget https://www.openssl.org/source/openssl-3.0.5.tar.gz
+tar -xf openssl-3.0.5.tar.gz
+cd openssl-3.0.5
+./config --prefix=/usr/local/openssl --openssldir=/usr/local/openssl
+make
+make install
+# The command below is needed when building Arrow with Gandiva, so that
+# the CentOS7 default Openssl version 1.0 is not used, but the newer
+# version of Openssl that as just been installed is used by CMake
+echo 'export OPENSSL_ROOT_DIR=/usr/local/openssl' >> ~/.bash_profile
+source  ~/.bash_profile
```

Comment:
How about just disabling Gandiva with `export ARROW_GANDIVA=OFF`?
FYI: Our RPM packages for CentOS 7 disable Gandiva.

Reply:
CMake does not find Openssl1.1.1 as the binary is called openssl11 but is in the same location as openssl.  Could try adding a soft link to a different location? Installing seems to be preferable to get the latest features.  Packaging Arrow and its dependencies as a software collection may be a helpful way to distribute it.

Reply:
If we want to support Gandiva on CentOS 7, we should improve our https://github.com/apache/arrow/blob/master/cpp/cmake_modules/FindOpenSSLAlt.cmake for the case.

Reply:
Improving that is better than disabling.

Reply:
Gandiva with Ruby causes a segmentation fault, part of the log is below:
```
/tmp/arrow-HEAD.jR5Hm/install/lib/libgandiva-glib.so.1000(g(ggandiva_function_signature_set_property+0xb2) [0x7fa76faf6f72]
/lib64/libgobject-2.0.so.0(0x7fa7787bd659) [0x7fa7787bd659]
/lib64/libgobject-2.0.so.0(g_object_new_valist+0x371) [0x7fa7787bf1b1]
/lib64/libgobject-2.0.so.0(g_object_new+0x99) [0x7fa7787bf4f9]
/tmp/arrow-HEAD.jR5Hm/install/lib/libgandiva-glib.so.1000(_Z35ggandiva_function_signature_new_rawPKN7gandiva17FunctionSignatureE+0x25) [0x7fa76faf76a5]
/tmp/arrow-HEAD.jR5Hm/install/lib/libgandiva-glib.so.1000(ggandiva_native_function_get_signatures+0x2c)
```
There are no probems with Gandiva and C++/Python though.

Reply:
Could you show failed test names?

Reply:
Further output is below, trying to determine cause of segmentation fault
```
/root/arrow/cpp/src/gandiva/cache.cc:50: Creating gandiva cache with capacity of 500
/root/arrow/cpp/src/gandiva/engine.cc:112: Detected CPU Name : znver2
/root/arrow/cpp/src/gandiva/engine.cc:113: Detected CPU Features: +sse2 -tsxldtrk +cx16 +sahf -tbm -avx512ifma +sha -gfni -fma4 -vpclmulqdq +prfchw +bmi2 -cldemote +fsgsbase -ptwrite -amx-tile -uintr +popcnt -widekl +aes -avx512bitalg -movdiri -xsaves -avx512er -avxvnni -avx512vnni -amx-bf16 -avx512vpopcntdq -pconfig +clwb -avx512f +xsavec +clzero -pku +mmx -lwp +rdpid -xop +rdseed -waitpkg -kl -movdir64b +sse4a -avx512bw +clflushopt +xsave -avx512vbmi2 +64bit -avx512vl -serialize -hreset -invpcid -avx512cd +avx -vaes -avx512bf16 +cx8 +fma -rtm +bmi -enqcmd +rdrnd -mwaitx +sse4.1 +sse4.2 +avx2 +fxsr +wbnoinvd +sse +lzcnt +pclmul -prefetchwt1 +f16c +ssse3 -sgx -shstk +cmov -avx512vbmi -amx-int8 +movbe -avx512vp2intersect +xsaveopt -avx512dq +adx -avx512pf +sse3
...../root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/gobject-introspection-4.0.3/lib/gobject-introspection/loader.rb:688: [BUG] Segmentation fault at 0xfffffffffffffff8
ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0036 p:---- s:0221 e:000220 CFUNC  :invoke
c:0035 p:0143 s:0215 e:000214 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/gobject-introspection-4.0.3/lib/gobject-introspection/loader.rb:688
c:0034 p:0024 s:0206 e:000205 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/gobject-introspection-4.0.3/lib/gobject-introspection/loader.rb:559
c:0033 p:0023 s:0199 e:000195 METHOD /root/arrow/c_glib/test/gandiva/test-function-registry.rb:30
c:0032 p:0158 s:0191 e:000190 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:871
c:0031 p:0004 s:0183 e:000182 BLOCK  /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:566
c:0030 p:0026 s:0180 e:000179 BLOCK  /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/fixture.rb:276
c:0029 p:0026 s:0176 e:000175 BLOCK  /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/fixture.rb:276
c:0028 p:0052 s:0172 e:000171 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/fixture.rb:257
c:0027 p:0009 s:0163 e:000162 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/fixture.rb:292
c:0026 p:0005 s:0158 E:000fe8 BLOCK  /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:564 [FINISH]
c:0025 p:---- s:0154 e:000153 CFUNC  :catch
c:0024 p:0047 s:0150 E:001bb0 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:563
c:0023 p:0018 s:0144 E:0023d8 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:124
c:0022 p:0082 s:0135 e:000134 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:53
c:0021 p:0018 s:0128 E:001838 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:124
c:0020 p:0082 s:0119 e:000118 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:53
c:0019 p:0018 s:0112 E:001a28 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:124
c:0018 p:0082 s:0103 e:000102 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:53
c:0017 p:0018 s:0096 E:001508 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:124
c:0016 p:0082 s:0087 e:000086 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:53
c:0015 p:0023 s:0080 E:002108 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:67
c:0014 p:0045 s:0075 e:000074 BLOCK  /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:45
c:0013 p:0086 s:0072 E:0001e8 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:102
c:0012 p:0015 s:0063 E:001928 BLOCK  /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:41 [FINISH]
c:0011 p:---- s:0059 e:000058 CFUNC  :catch
c:0010 p:0056 s:0055 E:0018f8 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:39
c:0009 p:0006 s:0048 e:000047 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunner.rb:40
c:0008 p:0019 s:0044 e:000043 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunner.rb:25
c:0007 p:0015 s:0040 e:000039 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnerutilities.rb:24
c:0006 p:0010 s:0034 e:000033 BLOCK  /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/autorunner.rb:458
c:0005 p:0026 s:0031 e:000030 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/autorunner.rb:514
c:0004 p:0139 s:0026 e:000025 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/autorunner.rb:457
c:0003 p:0059 s:0020 e:000019 METHOD /root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/autorunner.rb:66
c:0002 p:0476 s:0011 E:0006a0 EVAL   test/run-test.rb:118 [FINISH]
c:0001 p:0000 s:0003 E:0014b0 (none) [FINISH]

-- Ruby level backtrace information ----------------------------------------
test/run-test.rb:118:in `<main>'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/autorunner.rb:66:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/autorunner.rb:457:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/autorunner.rb:514:in `change_work_directory'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/autorunner.rb:458:in `block in run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnerutilities.rb:24:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunner.rb:25:in `start'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunner.rb:40:in `start_mediator'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:39:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:39:in `catch'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:41:in `block in run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:102:in `with_listener'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:45:in `block (2 levels) in run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/ui/testrunnermediator.rb:67:in `run_suite'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:53:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:124:in `run_test'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:53:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:124:in `run_test'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:53:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:124:in `run_test'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:53:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testsuite.rb:124:in `run_test'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:563:in `run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:563:in `catch'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:564:in `block in run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/fixture.rb:292:in `run_setup'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/fixture.rb:257:in `run_fixture'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/fixture.rb:276:in `block in create_fixtures_runner'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/fixture.rb:276:in `block in create_fixtures_runner'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:566:in `block (2 levels) in run'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/test-unit-3.5.5/lib/test/unit/testcase.rb:871:in `run_test'
/root/arrow/c_glib/test/gandiva/test-function-registry.rb:30:in `test_found'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/gobject-introspection-4.0.3/lib/gobject-introspection/loader.rb:559:in `signatures'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/gobject-introspection-4.0.3/lib/gobject-introspection/loader.rb:688:in `invoke'
/root/arrow/c_glib/vendor/bundle/ruby/3.0.0/gems/gobject-introspection-4.0.3/lib/gobject-introspection/loader.rb:688:in `invoke'
```